membership/src/main/scala/zio/membership/hyparview/plumtree/PlumTree.scala

```diff
@@ -0,0 +1,188 @@
+                    .send(sender, msg)
+                    .foldCauseM(
+//                      e => log.error(s"Failed replying to Graft message from $sender", e),
+                      logError,
+                      _ => logDebug(s"Replied to Graft message from $sender")
+                    ) *> PeerState.addToEagerPeers(sender).commit
+              }
+            case IHave(uuids) =>
+              def process(uuid: UUID) =
+                deduplication.contains(uuid).commit.flatMap {
+                  case true =>
+                    logDebug(s"Received IHave for already received message $uuid")
+                  case false =>
+                    onReceive(Right((uuid, sender)))
+                }
+              ZIO.foreach_(uuids)(process)
+            case UserMessage(payload) =>
+              userOut(payload) <* logDebug(s"Received userMessage from $sender")
+            case m @ Gossip(uuid, payload) =>
+              deduplication.add(uuid, m).commit.flatMap {
+                case true =>
+                  // first occurrence
+                  onReceive(Left(uuid)) *>
+                    (PeerState
+                      .addToEagerPeers(sender) *> PeerState.eagerPushPeers[T].zip(PeerState.lazyPushPeers[T])).commit
+                      .flatMap {
+                        case (eagerPeers, lazyPeers) =>
+                          val sendEager = ZIO.foreach_(eagerPeers.filterNot(_ == sender)) { peer =>
+                            PeerService
+                              .send(peer, m)
+                              .foldCauseM(
+//                                e => log.error(s"Failed sending message to $peer", e),
+                                logError,
+                                _ => logInfo(s"Sent Gossip to $peer")
+                              )
+                          }
+
+                          // no need to filter here as we ensured sender is in eager peers
+                          val sendLazy = ZIO.foreach_(lazyPeers.map((_, uuid)))(gossipOut.tupled)
+
+                          sendEager *> sendLazy *> userOut(payload)
+                      }
+                case false =>
+                  // seen before -> disconnect sender
+                  PeerService
+                    .send(sender, Prune)
+                    .foldCauseM(
+//                      e => logError(s"Failed to send prune message to $sender", e),
+                      logError,
+                      _ => logDebug(s"Sent prune message to $sender")
+                    )
+              }
+          }
+      }
+    layer >>>
+      ZLayer.fromManaged {
+        for {
+          env          <- ZManaged.environment[TRandom with PeerService[T] with Logging with PeerState[T]]
+          messages     <- BoundedTMap.make[UUID, Gossip](deduplicationBuffer).toManaged_
+          gossip       <- Queue.sliding[(T, UUID)](gossipBuffer).toManaged_
+          userMessages <- Queue.sliding[Chunk[Byte]](messagesBuffer).toManaged_
```

Comment:
should that be ` Queue.sliding[Chunk[Byte]](messagesBuffer).toManaged(_.shutdown)?

Reply:
Done. I think there was some reason why I didn't do it orginally, but I forgot 🤷‍♂